src/DataStreams/PushingToViewsBlockOutputStream.cpp

```diff
@@ -70,428 +178,502 @@ PushingToViewsBlockOutputStream::PushingToViewsBlockOutputStream(
-        auto sink = storage->write(query_ptr, storage->getInMemoryMetadataPtr(), getContext());
+        size_t num_views = views_data->views.size();
+        const Settings & settings = context->getSettingsRef();
+        if (settings.parallel_view_processing)
+            views_data->max_threads = settings.max_threads ? std::min(static_cast<size_t>(settings.max_threads), num_views) : num_views;
 
-        metadata_snapshot->check(sink->getPort().getHeader().getColumnsWithTypeAndName());
+        std::vector<Block> headers;
+        headers.reserve(num_views);
+        for (const auto & chain : chains)
+            headers.push_back(chain.getOutputHeader());
 
-        replicated_output = dynamic_cast<ReplicatedMergeTreeSink *>(sink.get());
-        output = std::make_shared<PushingToSinkBlockOutputStream>(std::move(sink));
-    }
-}
+        auto copying_data = std::make_shared<CopyingDataToViewsTransform>(storage_header, views_data);
+        auto finalizing_views = std::make_shared<FinalizingViewsTransform>(std::move(headers), views_data);
+        auto out = copying_data->getOutputs().begin();
+        auto in = finalizing_views->getInputs().begin();
 
-Block PushingToViewsBlockOutputStream::getHeader() const
-{
-    /// If we don't write directly to the destination
-    /// then expect that we're inserting with precalculated virtual columns
-    if (output)
-        return metadata_snapshot->getSampleBlock();
-    else
-        return metadata_snapshot->getSampleBlockWithVirtuals(storage->getVirtuals());
-}
+        size_t max_parallel_streams = 0;
 
-/// Auxiliary function to do the setup and teardown to run a view individually and collect its metrics inside the view ThreadStatus
-void inline runViewStage(ViewRuntimeData & view, const std::string & action, std::function<void()> stage)
-{
-    Stopwatch watch;
+        std::list<ProcessorPtr> processors;
 
-    auto * original_thread = current_thread;
-    SCOPE_EXIT({ current_thread = original_thread; });
+        for (auto & chain : chains)
+        {
+            max_parallel_streams += std::max<size_t>(chain.getNumThreads(), 1);
+            result_chain.attachResources(chain.detachResources());
+            connect(*out, chain.getInputPort());
+            connect(chain.getOutputPort(), *in);
+            ++in;
+            ++out;
+            processors.splice(processors.end(), Chain::getProcessors(std::move(chain)));
+        }
 
-    if (view.runtime_stats.thread_status)
-    {
-        /// Change thread context to store individual metrics per view. Once the work in done, go back to the original thread
-        view.runtime_stats.thread_status->resetPerformanceCountersLastUsage();
-        current_thread = view.runtime_stats.thread_status.get();
+        processors.emplace_front(std::move(copying_data));
+        processors.emplace_back(std::move(finalizing_views));
+        result_chain = Chain(std::move(processors));
+        result_chain.setNumThreads(max_parallel_streams);
```

Comment:
I'm not sure if I'm reading this correctly, but this seems to be ignoring the `settings.parallel_view_processing` set above right? It seems that now `views_data->max_threads` is only used to log, but not to control parallelism. Is that correct and intended?

Reply:
I have probably missed it.
Actually, the idea is that views could be processed always in parallel (or not, which depends only on number of executing threads). But probably the proper number of threads is not set.

Reply:
>I'm not sure if I'm reading this correctly, but this seems to be ignoring the settings.parallel_view_processing set above right?

It should be accounted here - https://github.com/ClickHouse/ClickHouse/blob/3663d23e4d80fc5f1fcb8ff1578f49b18319fd26/src/Interpreters/InterpreterInsertQuery.cpp#L410-L412

>But probably the proper number of threads is not set.

#29786